ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/PeerDiscoveryController.java

```diff
@@ -231,17 +231,17 @@ public void onMessage(final Packet packet, final DiscoveryPeer sender) {
             .ifPresent(
                 interaction -> {
                   // Extract the peers from the incoming packet.
-                  final List<DiscoveryPeer> neighbors =
+                  final List<Peer> neighbors =
                       packet
                           .getPacketData(NeighborsPacketData.class)
                           .map(NeighborsPacketData::getNodes)
                           .orElse(emptyList());
 
-                  for (final DiscoveryPeer neighbor : neighbors) {
+                  for (final Peer neighbor : neighbors) {
                     if (peerBlacklist.contains(neighbor) || peerTable.get(neighbor).isPresent()) {
                       continue;
                     }
-                    bond(neighbor, false);
+                    bond((DiscoveryPeer) neighbor, false);
```

Comment:
If we're going to have to cast things back to `DiscoveryPeer` we should just be honest and go back to using the `DiscoveryPeer` type all the way through.  It should only be declared `Peer` if it really can be any implementation of `Peer`.